sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/TableAsyncBatch.java

```diff
@@ -0,0 +1,385 @@
+     */
+    public synchronized List<BatchOperation> getOperations() {
+        return Collections.unmodifiableList(this.operations);
+    }
+
+    /**
+     * Executes all operations within The batch inside a transaction. When the call completes, either all operations in
+     * the batch will succeed, or if a failure occurs, all operations in the batch will be rolled back.
+     *
+     * @return A reactive result containing a list of sub-responses for each operation in the batch.
+     * @throws TableServiceErrorException if any operation within the batch fails. See the documentation for the client
+     *                                    methods in {@link TableAsyncClient} to understand the conditions that may
+     *                                    cause a given operation to fail.
+     * @throws IllegalStateException if no operations have been added to the batch.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public synchronized Mono<List<BatchOperationResponse>> submitTransaction() {
+        return submitTransactionWithResponse().flatMap(response -> Mono.justOrEmpty(response.getValue()));
+    }
+
+    /**
+     * Executes all operations within The batch inside a transaction. When the call completes, either all operations in
+     * the batch will succeed, or if a failure occurs, all operations in the batch will be rolled back.
+     *
+     * @return A reactive result containing the HTTP response produced for the batch itself. The response's value will
+     * contain a list of sub-responses for each operation in the batch.
+     * @throws TableServiceErrorException if any operation within the batch fails. See the documentation for the client
+     *                                    methods in {@link TableAsyncClient} to understand the conditions that may
+     *                                    cause a given operation to fail.
+     * @throws IllegalStateException if no operations have been added to the batch.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public synchronized Mono<Response<List<BatchOperationResponse>>> submitTransactionWithResponse() {
+        return withContext(this::submitTransactionWithResponse);
+    }
+
+    synchronized Mono<Response<List<BatchOperationResponse>>> submitTransactionWithResponse(Context context) {
+        this.frozen = true;
+        context = context == null ? Context.NONE : context;
+
+        if (operations.size() == 0) {
+            throw logger.logExceptionAsError(new IllegalStateException("A batch must contain at least one operation."));
+        }
+
+        final BatchRequestBody body = new BatchRequestBody();
+        Flux.fromIterable(operations)
+            .flatMapSequential(op -> op.prepareRequest(operationClient))
+            .zipWith(Flux.fromIterable(operations))
+            .doOnNext(pair -> body.addChangeOperation(new BatchSubRequest(pair.getT2(), pair.getT1())))
```

Comment:
doOnNext should be for side effects like logging. There are better operators to use if you want it to be part of your main scenario. 

In this case, would you be better using `operations.stream()....`? There aren't any async operations going on.

Reply:
The call to `op.prepareRequest` is async, it returns a `Mono<HttpRequest>`. Because of that, I'm not clear if I can replace `.doOnNext(...)` with `.toStream().forEach(...)`, but I'm far from an expert at Reactor. Maybe I will borrow a couple minutes of you or Alan's time to see how to write this better - I definitely felt a little unconfident about this part.

Reply:
I would lean towards using `collect` operations on `Flux`.

```java
Flux.fromIterable(operations)
    .flatMapSequential(op -> op.prepareRequest(operationClient).zipWith(Mono.just(op)))
    .collect(/* collect into body here */);
```

Reply:
It can also then be chained with the code below:

```java
/* collect reactive call */
   .map(body -> /* code below */);
```